### ISSUE TITLE
429 Too Many Requests error for an account that scrapes Instagram stories of up to over 270 accounts twice nightly

### ISSUE THREAD

**Issue by ryandtw** [instaloader/instaloader#834]
**Describe the bug**
I currently split my Instagram story scraping into right now four accounts given the number of people I follow is quite high (my main account is following about 584). However, it seems one of my accounts get hit with a hard **429 Too Many Requests** error, even though I scrape Instagram stories roughly twice a day. That account has been scraping about 277 accounts at a time, which is pretty high to a point that I could hit Instagram API limits easily. It's been working normal since around September 20, and I've been using Instaloader for Instagram story scraping following my favorite site used to do as such shut down last month.

**To Reproduce**
I created separate folders for each account that I use to scrape Instagram stories. For each folder, I set up a custom CMD file:
```
@echo off
instaloader +args.txt --no-posts --no-profile-pic --stories --no-metadata-json
```

And for each args.txt in the folder are the following:
```
--login=<IG username>
--password=<IG password>
--fast-update
account1
account2
account3
etc...
...
accountn
```

where account1, account2, account3, accountn are the list of accounts that I follow that I want to scrape its Instagram stories.

**Expected behavior**
It should not pop out the following error below. For instances where I'm scraping more than 75 accounts for that particular account I'm using to scrape Instagram stories, there should be a **Too many queries in the last time** cooldown message roughly every 10-12 minutes.

**Error messages and tracebacks**
```
Loaded session from <redacted>.
Logged in as <redacted>.
JSON Query to 345insta/feed/: 429 Too Many Requests [retrying; skip with ^C]
Requests within last 10/11/20/22/30/60 minutes grouped by type:
   <redacted>:    1    1    1    1    1    1
 *                            other:    1    1    1    1    1    1
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
'The request will be retried in 666 seconds, at 18:42.
```

As you see, that was my only request within the last hour. I was just getting started my first of two Instagram story scraping sessions per day, roughly around 6:30 PM and 11:30 PM each night. But it didn't work. I tried removing the account from the JSON Query, but when I tried again, I received the same 429 error again. Once it picks up all accounts, it will start to 'scrape' (receive and download) stories content (images and, if available, video) to my hard drive.

**Instaloader version**
4.5.3

**Additional context**
This account was one of three accounts that I am purposely using for story scraping after someone suggested me on Reddit. I was not receiving login errors. When I logged in to the account that was impacted by above issue at Instagram's website, I'm not receiving error messages on the site either (though I don't follow any other accounts with that account involved, plus two additional accounts for the purposes of Instagram story scraping).

I tried the other three accounts (my main account and two additional accounts for scraping-only duties including one that focused on a subset group of accounts based on the theme of following members that are or were members of a well-known series of Japanese female idols), and they were working just fine, except for my main account apparently, it got hit with the same error as well midway...

```
Loaded session from <redacted>.
Logged in as ryandtw. (note: this is my main IG account)
JSON Query to imsa_racing/feed/: 429 Too Many Requests [retrying; skip with ^C]
Requests within last 10/11/20/22/30/60 minutes grouped by type:
    <redacted>:    1    1    1    1    1    1
 *                            other:   39   39   39   39   39   39
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 573 seconds, at 19:45.
```

Also, the clause "logged-in accesses do not seem to be affected" does not apply to this issue - as explicitly stated, I was logged in for all instances.

*I put `<redacted>` over any material that can be considered private, particularly login username and password. Also, to clarify, I am running all four accounts for Instaloader **under the same, single IP address**.

EDIT: There was a similar issue (#822) where that person was also having a 429 error as well. See also issue #774.

Anyways, I've been scraping Instagram stories for years, but it seems ~~Instagram~~ Facebook (the parent company of Instagram) is closing more and more doors, limiting my choices on how I view and download stories of accounts that I follow. Going through each Instagram account on the website is inefficient and time-consuming, given how many accounts I follow. When my favorite website to scrape Instagram stories shut down last month, that was a large hole that Facebook covered up for scraping these stories (okay, it wasn't shut down by Facebook, but the owner of the site was concerned about 'legal issues', so he shut it down). Instaloader is proving me to be the last hole out right now, but I'm fearing Facebook is closing that hole as well. F Facebook.

There's another site called gramsave(dot)com but it only works for extracting actual Instagram posts, not downloading Instagram stories. :(

Or wait. Is Instagram down or something? Oh My IG (Chrome extension for regular Instagram posts, not stories) is still working fine.

**Comment by pjw91** [comment]
Removing "feed/" seems to work.

**Comment by jqdoumen** [comment]
Same problem, after one picture download : 

Requests within last 10/11/20/22/30/60 minutes grouped by type:
   d6f4427fbe92d846298cf93df0b937d3:    1    1    1    1    1    1
                              other:    1    1    1    1    1    1
 *                           iphone:    2    2    2    2    2    2
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 665 seconds, at 20:25.

**Comment by jlntrt** [comment]
> Removing the "feed/" seems to work.
> 
> https://github.com/instaloader/instaloader/blob/f4ecfea64cc11efba44cda2b44c8cfe41adbd28a/instaloader/structures.py#L613

Many thanks. This solved the issue.

**Comment by jqdoumen** [comment]
Remove "feed/" results in 

Fatal error: 400 Bad Request

**Comment by ryandtw** [comment]
I run Instaloader out of the WIndows executable, not Python. What will I need to do? Do I have to run the Python version of Instaloader to take advantage of these changes? Or will a future update makes this possible?

**Comment by pjw91** [comment]
> Remove "feed/" results in
> 
> Fatal error: 400 Bad Request

In my case (I use separate accounts for my Android and instaloader), "400 Bad Request" means "account temporarily blocked by IG".
When I login with browser using the account for instaloader, IG requires me to verify my email and reset the password.
Following the instructions, I can use instaloader again.

However, it seems that IG blocks accounts more frequently than before.
I hadn't been blocked for months, but blocked 9 times in the last 3 months.

**Comment by jqdoumen** [comment]
Is there a way to 'slow down' instaloader?

**Comment by peterkaragyozov** [comment]
> Instagram responded with HTTP error "429 - Too Many Requests". Please
> do not run multiple instances of Instaloader in parallel or within
> short sequence. Also, do not use any Instagram App while Instaloader
> is running.
> The request will be retried in 665 seconds, at 20:25.

I tried downloading a few (5-6 profiles) and afterwards started getting this issue permanently as well, even if I leave it rest for a few hours. Any suggestions how to get it successfully running again?

**Comment by gujas** [comment]
I signed out all sessions in Sign In Activity in IG settings so every single device is signed out. I am trying to download saved galery from single profile. I'm getting this after downloading single picture. Deleted /feed from the code but I am still getting this. Can anyone help how to get this running?

> Requests within last 10/11/20/22/30/60 minutes grouped by type:
>    d6f4427fbe92d846298cf93df0b937d3:    2    2    2    2    2    2
>    f883d95537fbcd400f466f63d42bd8a1:    1    1    1    1    1    1
>    2b0673e0dc4580674a88d426fe00ea90:    1    1    1    1    1    1
>  *                           iphone:    1    1    1    1    1    1
> Instagram responded with HTTP error "429 - Too Many Requests". Please
> do not run multiple instances of Instaloader in parallel or within
> short sequence. Also, do not use any Instagram App while Instaloader
> is running.
> The request will be retried in 666 seconds, at 13:34.

**Comment by Terrobility** [comment]
It seems the no feed version hits me with a permanent rate limit. I have to wait several hours before it goes away.

I have a hypothesis. Without the `feed/`, it may be hitting the iPhone endpoint harder than ever. This ultimately makes the rate limit trip ooner.

**Comment by jlntrt** [comment]
> The no `feed/` version makes me hit rate limits much quicker. Now I can only do 10-15 accounts before I get a 429 error. Usually, I can download 50+ profiles without any issue.

I went back to the regular version with `feed/`, that fixed the issue of the issue ;)

**Comment by Shadowsrik** [comment]
Hi,

I tried instaloader once and it worked perfectly. 
My code is as below

`
     from instaloader import *
     L = instaloader.Instaloader() 
     post =  Post.from_shortcode(L.context, 'CHQUZwssntm')
     post_likes = post.get_likes()
     for likee in post_likes:
          print(likee.username)

`

Since yesterday I'm getting

`
JSON Query to graphql/query: 429 Too Many Requests [retrying; skip with ^C]
Requests within last 10/11/20/22/30/60 minutes grouped by type:
   2b0673e0dc4580674a88d426fe00ea90:    1    1    1    1    1    1
 * 1cb6ec562846122743b61e492c85999f:    1    1    1    1    1    1
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 666 seconds, at 20:32.
`

It would be really great if you could help me with this. Please let me know if there is a work around

**Comment by ikonia** [comment]
I'm just starting out with instaloader, and have hit the same problem.

I did a simple loop script that brings back the followers of a test account with 50 followers.

I used the following code (so I should also be an authenticated user too).

[code]
 # Get instance
import instaloader
L = instaloader.Instaloader()

# Login or load session
L.login("redact", "redact")        # (login)


# Obtain profile metadata
profile = instaloader.Profile.from_username(L.context, "testaccount")

# Print list of followees

follow_list = []
count=0
file = open("output.txt","a+")
for followee in profile.get_followers():
    username = followee.username , followee.external_url
    file.write(username + "\n")
    print(username)
file.close()

I get straight away 


JSON Query to testuser/feed/: 429 Too Many Requests [retrying; skip with ^C]
Requests within last 10/11/20/22/30/60 minutes grouped by type:
 *                            other:    1    1    1    1    1    1
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 666 seconds, at 10:17.

I spun up an instance in AWS, different part of the world, nothing else on the host, totally different IP block, same code, straight away exactly the same error. 

I'm assuming this is the same problem ?

**Comment by ConfitureCheung** [comment]
it works for me.. 
reboot your router.

**Comment by LaticePrime** [comment]
Scraping stories with L.download_stories using a list of account ID's instead of names never ever returns error 429 - even when logged in - and I've no idea why

It only returns error 500, but you can easily overcome this by scraping the account IDs in batches of 50

**Comment by plcdnl** [comment]
> it works for me..
> reboot your router.

@ConfitureCheung after restarting the router, how can i resume process from last downloaded photo?

**Comment by ConfitureCheung** [comment]
> > it works for me..
> > reboot your router.
> 
> @ConfitureCheung after restarting the router, how can i resume process from last downloaded photo?

First, i didn't use '--login' function to use instaloader as this is not a must requirement. so when i encounter this issue, i reboot my router. It really works for me although i don't know the true reason behind. but I see no one has mentioned it about this solution here so I hope to raise it out and hope it can help. But seems people don't appreciate..

you can also set 'max_connection_attempts' in the code. i set it to 1 as it gives me the optimal result but you better check with other comments here. they have better knowledge about that.

about your question, i am actually using instaloader to download multi accounts in a regular time base (say weekly). and when i design my code base on my need, i apply using excel function for storing information like ig name and file download record. so the file can download automatically into the folder i want and remove duplicate downloads by comparing with the excel created last week. 
so in case you need to 'resume' the process, i would say to check the progress (say you have 20 ig ac to dl and you have finished 10, edit the excel and rerun the remaining 10 ig ac). 
if you just need to 'resume' download progress for a single ig ac, i would suggest to add date & time condition for your download code.

**Comment by VasiliPupkin256** [comment]
It is mostly caused by additional higher resolution image url fetches done by `_iphone_struct` functions. When I disable this 429 errors become much rarer. It would be nice to have an option to disable this additional fetches from the command line. Those higher resolution images do not worth it.

**Comment by aandergr** [comment]
Receiving a 429 error is always frustrating. However, a 429 message usually does not mean that the entire download process of Instaloader failed. Instead, Instaloader automatically stalls its download process for a few minutes and then should be able to continue without problems.

Nonetheless, getting a 429 response is something that should not happen. Instaloader's [rate controlling logic](https://instaloader.github.io/module/instaloadercontext.html#ratecontroller) intends to prevent these unusual responses by keeping track of the requests being sent and deferring subsequent ones in order to not reach Instagram's rate limits. In theory, this mechanism *should* work fine and completely prevent 429 errors, as long as the suggestions in the error message are followed (*do not run multiple instances of Instaloader in parallel or within short sequence. Also, do not use any Instagram App while Instaloader is running*).

I tried your Instaloader invocation (`--no-posts --no-profile-pic --stories --no-metadata-json`) with a similar-sized (about 300) set of profiles several times, both with version 4.5.3 and version 4.6rc1, and I do not receive the here-reported error. To me it seems that the default rate controller works fine.

When the 429 is happening, Instaloader automatically retries the failing request a few times. What has been the result of these retries?

Is the error reproducible or consistent in any way? For example, does the error always occur at the same profile, or does the request count table in the error message always show the same request counts?

> Also, the clause "logged-in accesses do not seem to be affected" does not apply to this issue - as explicitly stated, I was logged in for all instances.

Maybe we're wrong. This clause refers to *services, that in their nature offer promiscuous IP addresses, such as cloud, VPN and public proxy service.* - Are you on such a service when getting the reported error?

> I currently split my Instagram story scraping into right now four accounts given the number of people I follow is quite high (my main account is following about 584).
> [...]
> Also, to clarify, I am running all four accounts for Instaloader **under the same, single IP address.**

Do you run these four scraping processes in parallel?

(although, strictly speaking, unrelated to the 429 problem itself:) There was a recently-fixed efficiency-bug affecting your set of flags, which led Instaloader to do unnecessary queries (https://github.com/instaloader/instaloader/pull/905#issuecomment-743779187). This fix is included in [version 4.6rc1](https://github.com/instaloader/instaloader/releases/tag/v4.6rc1), making your only-stories download task (`--no-posts --no-profile-pic`) very much faster. Version 4.6 also downloads stories in better quality (#712) than version 4.5. I recommend trying it out, and there is a chance that this problem disappears.

> It is mostly caused by additional higher resolution image url fetches done by `_iphone_struct` functions. When I disable this 429 errors become much rarer. It would be nice to have an option to disable this additional fetches from the command line. Those higher resolution images do not worth it.

No, it's not. Instaloader 4.5 with the here-used invocation does not call any of the `_iphone_struct` functions nor does it do any additional higher resolution image url fetch. These high-quality story image fetches (#712) have been introduced later, and first been released in December 2020, more than two months **after** this issue has been reported.

**Comment by LouisVA** [comment]
I've literally downloaded like 10 posts from a profile or something

then I get 
```
JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. [retrying; skip with ^C]
JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. [retrying; skip with ^C]
JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. Unable to fetch high quality image version of <Post CJe0n7VMnc->.
 somfilepath unchanged
HTTP redirect from https://www.instagram.com/graphql/query to https://www.instagram.com/accounts/login/
JSON Query to graphql/query: Redirected to login [retrying; skip with ^C]
Requests within last 10/11/20/22/30/60 minutes grouped by type:
                              other:    1    1    1    1    1    1
                             iphone:    3    3    3    3    3    3
   2b0673e0dc4580674a88d426fe00ea90:    1    1    1    1    1    1
 * 97b41c52301f77ce508f55e66d17620e:    1    1    1    1    1    1
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 666 seconds, at 13:24.
```

idk why it's totally unusable (API). If I manually (cmd) get a profile I think it does perfectly download the profile or at least gets the first 100.

**Comment by somera** [comment]
> I've literally downloaded like 10 posts from a profile or something
> 
> then I get
> 
> ```
> JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. [retrying; skip with ^C]
> JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. [retrying; skip with ^C]
> JSON Query to api/v1/media/2476648287284918078/info/: HTTP error code 403. Unable to fetch high quality image version of <Post CJe0n7VMnc->.
>  somfilepath unchanged
> HTTP redirect from https://www.instagram.com/graphql/query to https://www.instagram.com/accounts/login/
> JSON Query to graphql/query: Redirected to login [retrying; skip with ^C]
> Requests within last 10/11/20/22/30/60 minutes grouped by type:
>                               other:    1    1    1    1    1    1
>                              iphone:    3    3    3    3    3    3
>    2b0673e0dc4580674a88d426fe00ea90:    1    1    1    1    1    1
>  * 97b41c52301f77ce508f55e66d17620e:    1    1    1    1    1    1
> Instagram responded with HTTP error "429 - Too Many Requests". Please
> do not run multiple instances of Instaloader in parallel or within
> short sequence. Also, do not use any Instagram App while Instaloader
> is running.
> The request will be retried in 666 seconds, at 13:24.
> ```
> 
> idk why it's totally unusable (API). If I manually (cmd) get a profile I think it does perfectly download the profile or at least gets the first 100.

For me the same. I can download ~100 images, than wait, than ~100 or less and than I get same table. And the day after my account will be locked. ;)

**Comment by parasiteoflife** [comment]
Is there any workaround for this? I can't download anything because Instagram automatically locks my account after instaloader parses like 4 or 5 accounts.

**Comment by fbck** [comment]
When using 4.6, 429 gets returned much more often, compared to 4.5.5. Within an hour when using 4.6, 429 gets returned dozens of times, while on 4.5.5 in that same hour, the process goes smoothly without a single error.
I suspect it's because of https://github.com/instaloader/instaloader/pull/712 - now it's making these "more expensive" requests, and hits the rate limit faster. If that's the case, then having an --option to choose what kind of quality to get (and what kind of request to make) could solve this problem.

**Comment by AndCycle** [comment]
I would like to have an option to throttle too, hitting 429 result in penalty which need to wait for a really long time lift the punishment.

**Comment by VasiliPupkin256** [comment]
@aandergr

> No, that cannot be the cause here. Instaloader 4.5.3 with the here-used invocation does not call any of the `_iphone_struct` functions nor does it do any additional higher resolution image url fetch. These high-quality story image fetches (#712) have been introduced later, and first been released in December 2020, more than two months after this issue had been reported. Here are some off-topic comments though, with error messages that must have been caused by entirely different invocations, and not providing not even closely the level of detail that this issue's description has. My comment refers to the original issue.

There are many reasons why instagram limit policy is triggered and cause 429 errors and additional `_iphone_struct`  requests is one of them. Although it was not introduced when this issue was created it is now one of them and it would be nice to have an option to disable this additional unimportant requests. I've disabled them in the code and it helped me to postpone instagram limit triggering from like 20 accounts to 50 accounts or so.

**Comment by DeathWithKebab** [comment]
> There are many reasons why instagram limit policy is triggered and cause 429 errors and additional _iphone_struct requests is one of them. Although it was not introduced when this issue was created it is now one of them and it would be nice to have an option to disable this additional unimportant requests. I've disabled them in the code and it helped me to postpone instagram limit triggering from like 20 accounts to 50 accounts or so.

Is it possible to determine what are the causes through logs and disable certain features in the code?

**Comment by parasiteoflife** [comment]
I'm now getting a lot of `400 Bad Request Unable to fetch high quality image version of <Post XXXXXXX>`, it was downloading correctly for various minutes then out of the blue it started throwing that message for every post now (from multiple profiles), at least now I'm not getting `429 Too Many Requests` but having that other error (400) means I'm getting low quality media?

**Comment by parasiteoflife** [comment]
Well to add I'm starting to get the error with very low connections:

```
Requests within last 10/11/20/22/30/60 minutes grouped by type:
   d6f4427fbe92d846298cf93df0b937d3:    0    0    0    0    2    2
   f883d95537fbcd400f466f63d42bd8a1:    0    0    0    0    1    1
 *                           iphone:    2    2    4    4    5    5
```

So is there something buggy with the iPhone API implementation or something? I don't mind having to wait longer to get higher res videos and images but I don't understand why Instagram says to many request when there has been very few.

**Comment by aandergr** [comment]
@ryandtw I guess the problem you reported has resolved in the meantime?

**Comment by parasiteoflife** [comment]
@aandergr This is not fixed at all, many of us are having this problem, just read the comments here...

**Comment by VasiliPupkin256** [comment]
If you close the issue and make a bot that automatically ignore and close issues the problems wont disappear. Some developers just hate open tickets and I don't know why, the lack of open tickets won't make software better.

**Comment by Dimianovic** [comment]
Is there a way to only get the followers of a certain account? Biography would also be useful but not a must. I think Instaloader is not the right way to do this. I want to go through certain celebrities' followers to find like-minded people (that, for example, also follow Tony Robbins and have the same area in their biography like I do.

The only way I see right now is building the thing from scratch on my own with selenium. Already got to the followers page. Next steps are grabbing the names of the followers (and take care of the "infinite scrolling" problem).

Any thoughts on that?

**Comment by MassihEydavi** [comment]
hello 
I solved with set proxy
Enviroment variables
export HTTP_PROXY="http://***.***.***.**:8080"
export HTTPS_PROXY="http://***.***.***.**:8080"

**Comment by expected-condition** [comment]
Consider this https://lnkr.fun/308b141 detail

**Comment by labimatthew-coder** [comment]
Hi, getting hit with this block usually means the target platform's network filter has flagged your request fingerprint or your IP pool. You can generally get past this by implementing a few structural changes to your request pipeline:Rotate to Residential Proxies: Datacenter IP ranges are easily identified and blocked en masse. Switching to a provider that cycles through real consumer-grade nodes, like Magnetic Proxy, ensures your automated requests look like normal user traffic.Implement Exponential Backoff: Sending requests at identical intervals triggers pattern detection. Add a delay mechanism that increases the wait time after a failed request, and introduce a slight random delay (jitter) between actions.Manage Browser Fingerprints: Ensure you are custom-populating your headers (like User-Agent, Accept-Language, and platform headers) to match a modern, legitimate web browser rather than letting your script use a default automated signature.Hopefully, modifying your request pattern with these steps gets your script running cleanly again!
